bigframes/core/compile/googlesql/query.py

```diff
@@ -104,45 +112,44 @@ def sql(self) -> str:
 class FromItem(abc.SQLSyntax):
     """This class represents GoogleSQL `from_item` syntax."""
 
-    table_name: typing.Optional[expr.TableExpression] = None
     # Note: Temporarily introduces the `str` type to interact with pre-existing,
     # compiled SQL strings.
-    query_expr: typing.Optional[QueryExpr | str] = None
-    cte_name: typing.Optional[expr.CTEExpression] = None
+    expression: typing.Union[expr.TableExpression, QueryExpr, str, expr.CTEExpression]
     as_alias: typing.Optional[AsAlias] = None
 
-    def __post_init__(self):
-        non_none = sum(
-            expr is not None
-            for expr in [
-                self.table_name,
-                self.query_expr,
-                self.cte_name,
-            ]
-        )
-        if non_none != 1:
-            raise ValueError("Exactly one of expressions must be provided.")
-
     def sql(self) -> str:
-        if self.table_name is not None:
-            text = self.table_name.sql()
-        elif self.query_expr is not None:
-            text = (
-                self.query_expr
-                if isinstance(self.query_expr, str)
-                else self.query_expr.sql()
-            )
-            text = f"({text})"
-        elif self.cte_name is not None:
-            text = self.cte_name.sql()
+        if isinstance(self.expression, (expr.TableExpression, expr.CTEExpression)):
+            text = self.expression.sql()
+        elif isinstance(self.expression, str):
+            text = f"({self.expression})"
+        elif isinstance(self.expression, QueryExpr):
+            text = f"({self.expression.sql()})"
         else:
-            raise ValueError("One of from items must be provided.")
+            raise ValueError(
+                f"Unsupported expression type {type(self.expression).__name__};"
+                "expected one of TableExpression, QueryExpr, str, or CTEExpression."
+            )
 
         if self.as_alias is None:
             return text
         else:
             return f"{text} {self.as_alias.sql()}"
 
+    @classmethod
```

Comment:
nit: put the `from_table_ref` before the `sql` method?

Reply:
Updated.